### DISCUSSION TITLE
package archive / old versions for renv

### DISCUSSION THREAD

**Discussion by pfuehrlich-pik** [r-universe-org/help#91] [Ideas]
At the moment only the most recent version of each package is available via r-universe. For tools like [renv](https://rstudio.github.io/renv/) old package versions are needed as well. Is it possible to setup a universe in a way that supports such a package archive or is this planned in the near future?

Thanks for the great work! Especially the binary builds have already helped us at [PIK](https://pik-piam.r-universe.dev). We would like to switch to r-universe entirely and get rid of our own package server, but we also want to use renv...

**Comment by jeroen** [comment]
Currently we don't plan to archive all the versions and binaries, but we do plan to support some other things, like making snapshots of a universe.

We also try to support renv by adding the upstream package git repo and hash in the pkg DESCRIPTION file, just like when the package is installed using `install_github()`. Hopefully renv will be able to use this information to restore old versions of r-universe packages by getting them directly from git(hub). But I haven't tested this myself yet.

**Reply by pfuehrlich-pik** [reply]
Unfortunately `renv` does not use the git repo and hash from the `DESCRIPTION` file, I just checked. If a packaged was installed from a "Repository" (in `renv` terms that's CRAN and r-universe but not GitHub) git repo and hash are not stored in `renv.lock`. `renv::restore` is using only the `renv.lock` to restore all packages. Here's a sample `renv.lock` with a package installed from r-universe and a package installed from GitHub:
```
{
  "R": {
    "Version": "4.1.1",
    "Repositories": [
      {
        "Name": "pikpiam",
        "URL": "https://pik-piam.r-universe.dev"
      },
      {
        "Name": "CRAN",
        "URL": "https://cran.rstudio.com"
      }
    ]
  },
  "Packages": {
    "gdxtools": {
      "Package": "gdxtools",
      "Version": "0.6.0",
      "Source": "GitHub",
      "RemoteType": "github",
      "RemoteHost": "api.github.com",
      "RemoteUsername": "lolow",
      "RemoteRepo": "gdxtools",
      "RemoteRef": "master",
      "RemoteSha": "94b3509ba341a8eb75543bdb5dee75fe2a8b6348"
    },
    "remulator": {
      "Package": "remulator",
      "Version": "1.18.1",
      "Source": "Repository"
    },
    "renv": {
      "Package": "renv",
      "Version": "0.14.0",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "30e5eba91b67f7f4d75d31de14bbfbdc"
    }
  }
}
```

**Reply by pfuehrlich-pik** [reply]
There a some issues on the `renv` repository about `r-universe` compatibility, so maybe someone is already working on it:
https://github.com/rstudio/renv/issues/789
https://github.com/rstudio/renv/issues/786
https://github.com/rstudio/renv/issues/774

**Reply by jeroen** [reply]
OK yes maybe I should talk to @kevinushey. I was hoping that because all R-universe packages contain the `RemoteUrl` and  `RemoteRef` and `RemoteSha` fields now in the DESCRIPTION, renv would naturally be able to restore them, just like when the package was installed from git directly. But maybe it's not that simple.

I think indeed the issue to track is https://github.com/rstudio/renv/issues/789
